pkg/cacheutil/memcached_client.go

```diff
@@ -0,0 +1,382 @@
+	config.applyDefaults()
+	if err := config.validate(); err != nil {
+		return nil, err
+	}
+
+	c := &memcachedClient{
+		logger:        logger,
+		config:        config,
+		client:        client,
+		selector:      selector,
+		provider:      provider,
+		asyncQueue:    make(chan func(), config.MaxAsyncBufferSize),
+		getMultiQueue: make(chan *memcachedGetMultiBatch),
+		stop:          make(chan struct{}, 1),
+	}
+
+	// As soon as the client is created it must ensure that memcached server
+	// addresses are resolved, so we're going to trigger an initial addresses
+	// resolution here.
+	if err := c.resolveAddrs(); err != nil {
+		return nil, err
+	}
+
+	go c.resolveAddrsLoop()
+
+	// Start a number of goroutines - processing async operations - equal
+	// to the max concurrency we have.
+	c.asyncWorkers.Add(c.config.MaxAsyncConcurrency)
+	for i := 0; i < c.config.MaxAsyncConcurrency; i++ {
+		go c.asyncQueueProcessLoop()
+	}
+
+	// Start a number of goroutines - processing get multi batch operations - equal
+	// to the max concurrency we have.
+	c.getMultiWorkers.Add(c.config.MaxGetMultiBatchConcurrency)
+	for i := 0; i < c.config.MaxGetMultiBatchConcurrency; i++ {
+		go c.getMultiQueueProcessLoop()
+	}
+
+	return c, nil
+}
+
+func (c *memcachedClient) Stop() {
+	close(c.stop)
+
+	// Wait until all workers have terminated.
+	c.asyncWorkers.Wait()
+	c.getMultiWorkers.Wait()
+}
+
+func (c *memcachedClient) SetAsync(key string, value []byte, ttl time.Duration) error {
+	return c.enqueueAsync(func() {
+		err := c.client.Set(&memcache.Item{
+			Key:        key,
+			Value:      value,
+			Expiration: int32(time.Now().Add(ttl).Unix()),
+		})
+
+		if err != nil {
+			level.Warn(c.logger).Log("msg", fmt.Sprintf("failed to store item with key %s to memcached", key), "err", err)
+		}
+	})
+}
+
+func (c *memcachedClient) GetMulti(keys []string) (map[string][]byte, error) {
+	batches, err := c.getMultiBatched(keys)
+	if err != nil {
+		if len(batches) == 0 {
+			return nil, err
+		}
+
+		// In case we have both results and an error, it means some batch requests
+		// failed and other succeeded. In this case we prefer to log it and move on,
+		// given returning from results from the cache is better than returning
+		// nothing.
+		level.Warn(c.logger).Log("msg", "failed to fetch some keys batches from memcached", "err", err)
+	}
+
+	hits := map[string][]byte{}
+	for _, items := range batches {
+		for key, item := range items {
+			hits[key] = item.Value
+		}
+	}
+
+	return hits, nil
+}
+
+func (c *memcachedClient) getMultiBatched(keys []string) ([]map[string]*memcache.Item, error) {
+	// Do not batch if the input keys are less then the max batch size.
+	if len(keys) <= c.config.MaxGetMultiBatchSize {
+		items, err := c.client.GetMulti(keys)
+		if err != nil {
+			return nil, err
+		}
+
+		return []map[string]*memcache.Item{items}, nil
+	}
+
+	// Split input keys into batches and schedule a job for it.
+	results := make(chan *memcachedGetMultiResult)
```

Comment:
why not size this based on numResults to keep from blocking the multi queue?

Reply:
Makes sense. Fixed.